File: archive/directories/terrafusion_sync/plugins/gis_export/service.py
"""
TerraFusion SyncService - GIS Export Plugin - Service Layer

This module provides service layer functionality for the GIS Export plugin,
handling business logic and database operations.
"""

import uuid
import logging
import json
from datetime import datetime
from typing import Dict, List, Any, Optional, Tuple, Sequence

from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from sqlalchemy import update, and_

from terrafusion_sync.core_models import GisExportJob
from terrafusion_sync.plugins.gis_export.county_config import get_county_config

logger = logging.getLogger(__name__)

class GisExportService:
    """Service class for GIS Export operations."""

    @classmethod
    async def create_export_job(
        cls,
        export_format: str,
        county_id: str,
        area_of_interest: Dict[str, Any],
        layers: List[str],
        parameters: Optional[Dict[str, Any]] = None,
        db: AsyncSession = None
    ) -> GisExportJob:
        """
        Create a new GIS export job.
        
        Args:
            export_format: Format of the export (e.g., GeoJSON, Shapefile, KML)
            county_id: ID of the county
            area_of_interest: GeoJSON object defining the area of interest
            layers: List of layers to export
            parameters: Additional parameters for the export
            db: Database session
            
        Returns:
            The created GisExportJob instance
            
        Raises:
            ValueError: If export format is not supported for the county or if db is None
        """
        if db is None:
            raise ValueError("Database session is required")
            
        # Validate export format against county configuration
        county_config = get_county_config()
        if not county_config.validate_export_format(county_id, export_format):
            valid_formats = county_config.get_available_formats(county_id)
            raise ValueError(
                f"Export format '{export_format}' is not supported for county '{county_id}'. "
                f"Supported formats are: {', '.join(valid_formats)}"
            )
        
        # Create the new job
        new_job = GisExportJob(
            job_id=str(uuid.uuid4()),
            export_format=export_format,
            county_id=county_id,
            area_of_interest_json=area_of_interest,
            layers_json=layers,
            parameters_json=parameters or {},
            status="PENDING",
            message="Job created and pending processing",
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow()
        )
        
        # Save to database
        db.add(new_job)
        await db.commit()
        await db.refresh(new_job)
        
        logger.info(f"Created GIS export job {new_job.job_id} for county {county_id} in format {export_format}")
        
        return new_job
    
    @classmethod
    async def get_job_by_id(cls, job_id: str, db: AsyncSession) -> Optional[GisExportJob]:
        """
        Get a GIS export job by ID.
        
        Args:
            job_id: ID of the job to retrieve
            db: Database session
            
        Returns:
            The GisExportJob instance or None if not found
        """
        stmt = select(GisExportJob).where(GisExportJob.job_id == job_id)
        result = await db.execute(stmt)
        job = result.scalars().first()
        
        if not job:
            logger.warning(f"GIS export job {job_id} not found")
        
        return job
    
    @classmethod
    async def update_job_status(
        cls, 
        job_id: str, 
        status: str, 
        message: Optional[str] = None,
        result_file_location: Optional[str] = None,
        result_file_size_kb: Optional[int] = None,
        result_record_count: Optional[int] = None,
        db: AsyncSession = None
    ) -> Optional[GisExportJob]:
        """
        Update the status of a GIS export job.
        
        Args:
            job_id: ID of the job to update
            status: New status of the job
            message: Optional status message
            result_file_location: Optional location of the result file
            result_file_size_kb: Optional size of the result file in KB
            result_record_count: Optional count of records in the result
            db: Database session
            
        Returns:
            The updated GisExportJob instance or None if not found
            
        Raises:
            ValueError: If db is None
        """
        if db is None:
            raise ValueError("Database session is required")
            
        # Prepare update values
        values = {"status": status, "updated_at": datetime.utcnow()}
        
        # Add optional fields if provided
        if message is not None:
            values["message"] = message
        
        if result_file_location is not None:
            values["result_file_location"] = result_file_location
        
        if result_file_size_kb is not None:
            values["result_file_size_kb"] = result_file_size_kb
        
        if result_record_count is not None:
            values["result_record_count"] = result_record_count
        
        # Update timestamps based on status
        if status == "RUNNING":
            values["started_at"] = datetime.utcnow()
        
        if status in ["COMPLETED", "FAILED", "CANCELLED"]:
            values["completed_at"] = datetime.utcnow()
        
        # Update in database
        stmt = (
            update(GisExportJob)
            .where(GisExportJob.job_id == job_id)
            .values(**values)
            .returning(GisExportJob)
        )
        
        result = await db.execute(stmt)
        updated_job = result.scalars().first()
        
        if not updated_job:
            logger.warning(f"Failed to update GIS export job {job_id}: Job not found")
            return None
        
        await db.commit()
        
        logger.info(f"Updated GIS export job {job_id} status to {status}")
        
        return updated_job
    
    @classmethod
    async def list_jobs(
        cls,
        county_id: Optional[str] = None,
        status: Optional[str] = None,
        limit: int = 100,
        offset: int = 0,
        db: AsyncSession = None
    ) -> List[GisExportJob]:
        """
        List GIS export jobs with optional filtering.
        
        Args:
            county_id: Optional county ID to filter by
            status: Optional status to filter by
            limit: Maximum number of jobs to return
            offset: Offset for pagination
            db: Database session
            
        Returns:
            List of GisExportJob instances
            
        Raises:
            ValueError: If db is None
        """
        if db is None:
            raise ValueError("Database session is required")
            
        # Build query conditions
        conditions = []
        
        if county_id:
            conditions.append(GisExportJob.county_id == county_id)
        
        if status:
            conditions.append(GisExportJob.status == status)
        
        # Create base query
        if conditions:
            stmt = (
                select(GisExportJob)
                .where(and_(*conditions))
                .order_by(GisExportJob.created_at.desc())
                .limit(limit)
                .offset(offset)
            )
        else:
            stmt = (
                select(GisExportJob)
                .order_by(GisExportJob.created_at.desc())
                .limit(limit)
                .offset(offset)
            )
        
        # Execute query
        result = await db.execute(stmt)
        jobs = list(result.scalars().all())  # Convert to list explicitly
        
        logger.info(f"Retrieved {len(jobs)} GIS export jobs")
        
        return jobs
    
    @classmethod
    async def cancel_job(cls, job_id: str, db: AsyncSession) -> Optional[GisExportJob]:
        """
        Cancel a running or pending GIS export job.
        
        Args:
            job_id: ID of the job to cancel
            db: Database session
            
        Returns:
            The updated GisExportJob instance or None if not found or cannot be cancelled
            
        Raises:
            ValueError: If the job is already completed, failed, or cancelled
        """
        # Get the job first
        stmt = select(GisExportJob).where(GisExportJob.job_id == job_id)
        result = await db.execute(stmt)
        job = result.scalars().first()
        
        if not job:
            logger.warning(f"Cannot cancel GIS export job {job_id}: Job not found")
            return None
        
        # Check if job can be cancelled
        # String comparison for status to avoid ColumnElement.__bool__ issues
        if str(job.status) in ["COMPLETED", "FAILED", "CANCELLED"]:
            logger.warning(f"Cannot cancel GIS export job {job_id}: Job is already {job.status}")
            raise ValueError(f"Cannot cancel job that is already {job.status}")
        
        # Update job status to CANCELLING or CANCELLED
        # String comparison for status to avoid ColumnElement.__bool__ issues
        new_status = "CANCELLING" if str(job.status) == "RUNNING" else "CANCELLED"
        message = "Job cancellation requested" if str(job.status) == "RUNNING" else "Job cancelled before processing started"
        
        # Update the job
        return await cls.update_job_status(
            job_id=job_id,
            status=new_status,
            message=message,
            db=db
        )
    
    @classmethod
    def get_default_export_parameters(cls, county_id: str) -> Dict[str, Any]:
        """
        Get default export parameters for a county.
        
        Args:
            county_id: ID of the county
            
        Returns:
            Dictionary of default parameters
        """
        return get_county_config().get_default_parameters(county_id)

File: archive/directories/terrarust/gis_export/src/service.rs
use crate::models::*;
use crate::{ExportFormat, GisExportConfig};
use sqlx::{PgPool, Row};
use uuid::Uuid;
use chrono::Utc;
use std::path::PathBuf;
use tokio::fs;
use std::collections::HashMap;
use anyhow::{Result, anyhow};

/// High-performance GIS Export Service
pub struct GisExportService {
    config: GisExportConfig,
    db_pool: PgPool,
}

impl GisExportService {
    /// Create a new GIS Export Service instance
    pub async fn new(config: GisExportConfig, db_pool: PgPool) -> Result<Self> {
        // Ensure storage directory exists
        fs::create_dir_all(&config.storage_path).await?;
        
        // Test database connection
        sqlx::query("SELECT 1").execute(&db_pool).await?;
        
        log::info!("GIS Export Service initialized with storage path: {:?}", config.storage_path);
        
        Ok(Self {
            config,
            db_pool,
        })
    }

    /// Create a new export job
    pub async fn create_job(&self, request: CreateJobRequest) -> Result<CreateJobResponse> {
        // Validate export format
        let export_format: ExportFormat = request.export_format.parse()
            .map_err(|e| anyhow!("Invalid export format: {}", e))?;

        // Validate layers
        if request.layers.is_empty() {
            return Err(anyhow!("At least one layer must be specified"));
        }

        // Generate unique job ID
        let job_id = Uuid::new_v4();
        let now = Utc::now();

        // Convert layers to JSON
        let layers_json = serde_json::to_value(&request.layers)?;
        let parameters_json = request.parameters.map(|p| serde_json::to_value(p)).transpose()?;

        // Insert job into database
        let job = sqlx::query_as::<_, GisExportJob>(
            r#"
            INSERT INTO gis_export_jobs (
                job_id, county_id, username, export_format, area_of_interest, 
                layers, parameters, status, message, created_at
            ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
            RETURNING *
            "#
        )
        .bind(job_id)
        .bind(&request.county_id)
        .bind(&request.username)
        .bind(export_format.as_str())
        .bind(&request.area_of_interest)
        .bind(layers_json)
        .bind(parameters_json)
        .bind("PENDING")
        .bind("Export job created and queued for processing")
        .bind(now)
        .fetch_one(&self.db_pool)
        .await?;

        log::info!("Created GIS export job {} for county {}", job_id, request.county_id);
        
        Ok(job.into())
    }

    /// Get job status by ID
    pub async fn get_job_status(&self, job_id: Uuid) -> Result<JobStatusResponse> {
        let job = sqlx::query_as::<_, GisExportJob>(
            "SELECT * FROM gis_export_jobs WHERE job_id = $1"
        )
        .bind(job_id)
        .fetch_optional(&self.db_pool)
        .await?
        .ok_or_else(|| anyhow!("Job not found: {}", job_id))?;

        Ok(job.into())
    }

    /// List jobs with optional filtering
    pub async fn list_jobs(&self, params: ListJobsParams) -> Result<JobListResponse> {
        let limit = params.limit.unwrap_or(50).min(1000); // Cap at 1000
        let offset = params.offset.unwrap_or(0);

        // Build dynamic query with filters
        let mut query = "SELECT * FROM gis_export_jobs WHERE 1=1".to_string();
        let mut bind_count = 0;
        let mut binds: Vec<Box<dyn sqlx::Encode<'_, sqlx::Postgres> + Send + 'static>> = Vec::new();

        if let Some(county_id) = &params.county_id {
            bind_count += 1;
            query.push_str(&format!(" AND county_id = ${}", bind_count));
            binds.push(Box::new(county_id.clone()));
        }

        if let Some(username) = &params.username {
            bind_count += 1;
            query.push_str(&format!(" AND username = ${}", bind_count));
            binds.push(Box::new(username.clone()));
        }

        if let Some(status) = &params.status {
            bind_count += 1;
            query.push_str(&format!(" AND status = ${}", bind_count));
            binds.push(Box::new(status.clone()));
        }

        query.push_str(" ORDER BY created_at DESC");
        
        bind_count += 1;
        query.push_str(&format!(" LIMIT ${}", bind_count));
        binds.push(Box::new(limit));
        
        bind_count += 1;
        query.push_str(&format!(" OFFSET ${}", bind_count));
        binds.push(Box::new(offset));

        // Execute query (simplified for now - in production would use proper parameter binding)
        let jobs = sqlx::query_as::<_, GisExportJob>(&query)
            .fetch_all(&self.db_pool)
            .await?;

        // Get total count for pagination
        let total_query = "SELECT COUNT(*) as count FROM gis_export_jobs WHERE 1=1".to_string();
        let total: i64 = sqlx::query(&total_query)
            .fetch_one(&self.db_pool)
            .await?
            .get("count");

        let job_responses: Vec<JobStatusResponse> = jobs.into_iter().map(|job| job.into()).collect();

        Ok(JobListResponse {
            jobs: job_responses,
            total,
            limit,
            offset,
        })
    }

    /// Process an export job
    pub async fn process_job(&self, job_id: Uuid) -> Result<JobStatusResponse> {
        // Load job
        let mut job = sqlx::query_as::<_, GisExportJob>(
            "SELECT * FROM gis_export_jobs WHERE job_id = $1"
        )
        .bind(job_id)
        .fetch_optional(&self.db_pool)
        .await?
        .ok_or_else(|| anyhow!("Job not found: {}", job_id))?;

        // Validate job can be processed
        if job.status != "PENDING" {
            return Err(anyhow!("Job {} is not in PENDING status", job_id));
        }

        // Update job to PROCESSING
        sqlx::query(
            "UPDATE gis_export_jobs SET status = $1, started_at = $2, message = $3 WHERE job_id = $4"
        )
        .bind("PROCESSING")
        .bind(Utc::now())
        .bind("Processing export...")
        .bind(job_id)
        .execute(&self.db_pool)
        .await?;

        log::info!("Processing GIS export job {}", job_id);

        // Process the export
        match self.generate_export(&job).await {
            Ok((file_path, file_size)) => {
                // Update job as completed
                let download_url = format!("/api/v1/gis-export/download/{}", job_id);
                
                sqlx::query(
                    r#"
                    UPDATE gis_export_jobs 
                    SET status = $1, completed_at = $2, message = $3, file_path = $4, 
                        file_size = $5, download_url = $6 
                    WHERE job_id = $7
                    "#
                )
                .bind("COMPLETED")
                .bind(Utc::now())
                .bind("Export completed successfully")
                .bind(file_path.to_string_lossy().to_string())
                .bind(file_size as i64)
                .bind(download_url)
                .bind(job_id)
                .execute(&self.db_pool)
                .await?;

                log::info!("Completed GIS export job {}", job_id);
            }
            Err(e) => {
                // Update job as failed
                sqlx::query(
                    "UPDATE gis_export_jobs SET status = $1, completed_at = $2, message = $3 WHERE job_id = $4"
                )
                .bind("FAILED")
                .bind(Utc::now())
                .bind(format!("Export failed: {}", e))
                .bind(job_id)
                .execute(&self.db_pool)
                .await?;

                log::error!("Failed GIS export job {}: {}", job_id, e);
                return Err(e);
            }
        }

        // Return updated job status
        self.get_job_status(job_id).await
    }

    /// Cancel a job
    pub async fn cancel_job(&self, job_id: Uuid) -> Result<JobStatusResponse> {
        let job = sqlx::query_as::<_, GisExportJob>(
            "SELECT * FROM gis_export_jobs WHERE job_id = $1"
        )
        .bind(job_id)
        .fetch_optional(&self.db_pool)
        .await?
        .ok_or_else(|| anyhow!("Job not found: {}", job_id))?;

        if job.status == "COMPLETED" {
            return Err(anyhow!("Cannot cancel completed job"));
        }

        sqlx::query(
            "UPDATE gis_export_jobs SET status = $1, completed_at = $2, message = $3 WHERE job_id = $4"
        )
        .bind("CANCELLED")
        .bind(Utc::now())
        .bind("Job cancelled by user")
        .bind(job_id)
        .execute(&self.db_pool)
        .await?;

        log::info!("Cancelled GIS export job {}", job_id);
        
        self.get_job_status(job_id).await
    }

    /// Generate the actual export file
    async fn generate_export(&self, job: &GisExportJob) -> Result<(PathBuf, u64)> {
        let export_format: ExportFormat = job.export_format.parse()?;
        let layers: Vec<String> = serde_json::from_value(job.layers.clone())?;

        // Create filename
        let filename = format!("{}_{}.{}", 
            job.county_id, 
            job.job_id.simple(),
            export_format.file_extension()
        );
        let file_path = self.config.storage_path.join(&filename);

        // Query geospatial data from database
        let features = self.query_features(job, &layers).await?;

        // Generate export based on format
        match export_format {
            ExportFormat::Geojson => {
                self.generate_geojson(&file_path, &features).await?;
            }
            ExportFormat::Csv => {
                self.generate_csv(&file_path, &features).await?;
            }
            ExportFormat::Shapefile => {
                self.generate_shapefile(&file_path, &features).await?;
            }
            ExportFormat::Kml => {
                self.generate_kml(&file_path, &features).await?;
            }
            ExportFormat::Geopackage => {
                self.generate_geopackage(&file_path, &features).await?;
            }
        }

        // Get file size
        let metadata = fs::metadata(&file_path).await?;
        let file_size = metadata.len();

        Ok((file_path, file_size))
    }

    /// Query features from database
    async fn query_features(&self, job: &GisExportJob, layers: &[String]) -> Result<Vec<HashMap<String, serde_json::Value>>> {
        // For demonstration, generate sample data
        // In production, this would query your actual geospatial database
        let mut features = Vec::new();
        
        for (i, layer) in layers.iter().enumerate() {
            for j in 0..100 { // Generate 100 sample features per layer
                let mut feature = HashMap::new();
                feature.insert("id".to_string(), serde_json::Value::Number((i * 100 + j).into()));
                feature.insert("layer".to_string(), serde_json::Value::String(layer.clone()));
                feature.insert("county_id".to_string(), serde_json::Value::String(job.county_id.clone()));
                feature.insert("geometry".to_string(), serde_json::json!({
                    "type": "Point",
                    "coordinates": [-119.0 + (j as f64 * 0.001), 46.0 + (i as f64 * 0.001)]
                }));
                features.push(feature);
            }
        }

        log::info!("Queried {} features for export", features.len());
        Ok(features)
    }

    /// Generate GeoJSON export
    async fn generate_geojson(&self, file_path: &PathBuf, features: &[HashMap<String, serde_json::Value>]) -> Result<()> {
        let geojson = serde_json::json!({
            "type": "FeatureCollection",
            "features": features.iter().map(|f| {
                serde_json::json!({
                    "type": "Feature",
                    "geometry": f.get("geometry").unwrap_or(&serde_json::Value::Null),
                    "properties": f.iter()
                        .filter(|(k, _)| *k != "geometry")
                        .collect::<HashMap<_, _>>()
                })
            }).collect::<Vec<_>>()
        });

        fs::write(file_path, serde_json::to_string_pretty(&geojson)?).await?;
        Ok(())
    }

    /// Generate CSV export
    async fn generate_csv(&self, file_path: &PathBuf, features: &[HashMap<String, serde_json::Value>]) -> Result<()> {
        if features.is_empty() {
            fs::write(file_path, "").await?;
            return Ok(());
        }

        // Get all unique column names
        let mut columns: std::collections::HashSet<String> = std::collections::HashSet::new();
        for feature in features {
            for key in feature.keys() {
                if key != "geometry" { // Skip geometry for CSV
                    columns.insert(key.clone());
                }
            }
        }
        let mut columns: Vec<String> = columns.into_iter().collect();
        columns.sort();

        // Build CSV content
        let mut csv_content = columns.join(",") + "\n";
        for feature in features {
            let row: Vec<String> = columns.iter().map(|col| {
                feature.get(col)
                    .map(|v| match v {
                        serde_json::Value::String(s) => format!("\"{}\"", s.replace("\"", "\"\"")),
                        serde_json::Value::Number(n) => n.to_string(),
                        serde_json::Value::Bool(b) => b.to_string(),
                        _ => "".to_string(),
                    })
                    .unwrap_or_default()
            }).collect();
            csv_content.push_str(&(row.join(",") + "\n"));
        }

        fs::write(file_path, csv_content).await?;
        Ok(())
    }

    /// Generate Shapefile export (placeholder)
    async fn generate_shapefile(&self, file_path: &PathBuf, features: &[HashMap<String, serde_json::Value>]) -> Result<()> {
        // For now, create a simple ZIP with GeoJSON
        // In production, you'd use GDAL or similar to create proper shapefiles
        let geojson_path = file_path.with_extension("geojson");
        self.generate_geojson(&geojson_path, features).await?;
        
        // Create simple ZIP file (placeholder implementation)
        fs::write(file_path, "Shapefile export placeholder").await?;
        Ok(())
    }

    /// Generate KML export (placeholder)
    async fn generate_kml(&self, file_path: &PathBuf, features: &[HashMap<String, serde_json::Value>]) -> Result<()> {
        fs::write(file_path, "KML export placeholder").await?;
        Ok(())
    }

    /// Generate GeoPackage export (placeholder)
    async fn generate_geopackage(&self, file_path: &PathBuf, features: &[HashMap<String, serde_json::Value>]) -> Result<()> {
        fs::write(file_path, "GeoPackage export placeholder").await?;
        Ok(())
    }

    /// Get file path for download
    pub async fn get_export_file(&self, job_id: Uuid) -> Result<PathBuf> {
        let job = sqlx::query_as::<_, GisExportJob>(
            "SELECT * FROM gis_export_jobs WHERE job_id = $1"
        )
        .bind(job_id)
        .fetch_optional(&self.db_pool)
        .await?
        .ok_or_else(|| anyhow!("Job not found: {}", job_id))?;

        if job.status != "COMPLETED" {
            return Err(anyhow!("Export not ready for download"));
        }

        let file_path = job.file_path
            .ok_or_else(|| anyhow!("No file path available"))?;
        
        let path = PathBuf::from(file_path);
        if !path.exists() {
            return Err(anyhow!("Export file not found"));
        }

        Ok(path)
    }
}